File: comprehensive_merge_script.py
#!/usr/bin/env python3
"""
Comprehensive merge script to handle all unmerged branches
Resolves conflicts by preferring main branch version and cleaning up conflicts
"""

import os
import subprocess
import sys
import re
from pathlib import Path

def run_command(cmd, cwd=None):
    """Run a command and return the result"""
    try:
        result = subprocess.run(cmd, shell=True, cwd=cwd, capture_output=True, text=True)
        return result.returncode == 0, result.stdout, result.stderr
    except Exception as e:
        return False, "", str(e)

def get_unmerged_branches():
    """Get list of branches that haven't been merged to main"""
    success, output, _ = run_command("git branch -r --no-merged main")
    if not success:
        return []
    
    branches = []
    for line in output.strip().split('\n'):
        if line.strip() and 'origin/' in line:
            branch = line.strip().replace('origin/', '')
            # Filter out backup branches and focus on feature branches
            if not branch.startswith('backup-main') and not branch.startswith('aggressive-merge-backup'):
                branches.append(branch)
    return branches

def resolve_conflicts_in_file(file_path):
    """Resolve conflicts in a single file by choosing main branch version"""
    if not os.path.exists(file_path):
        return True
    
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # Check if file has conflict markers
        if '<<<<<<< HEAD' not in content:
            return True
        
        # Split by conflict markers and keep main branch version
        lines = content.split('\n')
        resolved_lines = []
        in_conflict = False
        keep_main = True
        
        for line in lines:
            if line.strip() == '<<<<<<< HEAD':
                in_conflict = True
                keep_main = True
                continue
            elif line.strip() == '=======':
                keep_main = False
                continue
            elif line.strip().startswith('>>>>>>>'):
                in_conflict = False
                keep_main = True
                continue
            elif in_conflict and not keep_main:
                # Skip lines from the other branch
                continue
            else:
                resolved_lines.append(line)
        
        # Write resolved content
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write('\n'.join(resolved_lines))
        
        return True
    except Exception as e:
        print(f"Error resolving conflicts in {file_path}: {e}")
        return False

def resolve_all_conflicts():
    """Resolve all merge conflicts in the current merge state"""
    success, output, _ = run_command("git status --porcelain")
    if not success:
        return False
    
    conflicted_files = []
    for line in output.strip().split('\n'):
        if line.strip().startswith('UU') or line.strip().startswith('AA') or line.strip().startswith('DD'):
            file_path = line[3:].strip()
            conflicted_files.append(file_path)
    
    print(f"Found {len(conflicted_files)} conflicted files")
    
    for file_path in conflicted_files:
        print(f"Resolving conflicts in: {file_path}")
        if not resolve_conflicts_in_file(file_path):
            print(f"Failed to resolve conflicts in: {file_path}")
            return False
    
    return True

def merge_branch(branch_name):
    """Merge a specific branch with conflict resolution"""
    print(f"\n=== Merging branch: {branch_name} ===")
    
    # Start merge
    success, output, error = run_command(f"git merge origin/{branch_name}")
    
    if success:
        print(f"Successfully merged {branch_name}")
        return True
    
    # Check if there are conflicts
    if "CONFLICT" in output or "CONFLICT" in error:
        print(f"Conflicts detected in {branch_name}, resolving...")
        
        # Resolve conflicts
        if resolve_all_conflicts():
            # Add resolved files
            run_command("git add .")
            
            # Commit the merge
            success, _, _ = run_command(f"git commit -m 'Resolve merge conflicts for {branch_name}'")
            if success:
                print(f"Successfully resolved conflicts and merged {branch_name}")
                return True
            else:
                print(f"Failed to commit resolved conflicts for {branch_name}")
                return False
        else:
            print(f"Failed to resolve conflicts for {branch_name}")
            return False
    else:
        print(f"Merge failed for {branch_name}: {error}")
        return False

def main():
    """Main function to merge all unmerged branches"""
    print("Starting comprehensive merge process...")
    
    # Get list of unmerged branches
    branches = get_unmerged_branches()
    print(f"Found {len(branches)} unmerged branches")
    
    if not branches:
        print("No branches to merge")
        return
    
    # Filter to prioritize important branches
    priority_branches = []
    other_branches = []
    
    for branch in branches:
        if 'cursor/fix-errors-and-merge-to-main' in branch or 'codex/' in branch or 'add-' in branch or 'ai-' in branch:
            priority_branches.append(branch)
        else:
            other_branches.append(branch)
    
    # Merge priority branches first
    all_branches = priority_branches + other_branches
    print(f"Priority branches: {len(priority_branches)}")
    print(f"Other branches: {len(other_branches)}")
    
    # Merge each branch
    successful_merges = 0
    failed_merges = 0
    skipped_branches = 0
    
    for i, branch in enumerate(all_branches, 1):
        print(f"\n[{i}/{len(all_branches)}] Processing branch: {branch}")
        
        # Check if branch has commits ahead of main
        success, output, _ = run_command(f"git log --oneline main..origin/{branch}")
        if not success or not output.strip():
            print(f"Branch {branch} has no new commits, skipping")
            skipped_branches += 1
            continue
        
        if merge_branch(branch):
            successful_merges += 1
        else:
            failed_merges += 1
            # Abort current merge if it failed
            run_command("git merge --abort")
    
    print(f"\n=== Merge Summary ===")
    print(f"Successful merges: {successful_merges}")
    print(f"Failed merges: {failed_merges}")
    print(f"Skipped branches: {skipped_branches}")
    
    if successful_merges > 0:
        print("Pushing changes to remote...")
        run_command("git push origin main")
        print("Changes pushed successfully!")

if __name__ == "__main__":
    main()

File: comprehensive-merge-script.js
#!/usr/bin/env node
/**
 * Comprehensive Merge Script - Merges all unmerged branches into main
 * Handles merge conflicts systematically and provides detailed reporting
 */

import { execSync } from 'child_process';
import fs from 'fs';

// Function to safely execute git commands
function safeGitCommand(command, description) {
  try {
    console.log(`→ ${description}: ${command}`);
    const result = execSync(command, { 
      encoding: 'utf8', 
      stdio: 'pipe',
      timeout: 30000 // 30 second timeout
    });
    return { success: true, result };
  } catch (error) {
    console.log(`⚠️  ${description} failed: ${error.message}`);
    return { success: false, error: error.message };
  }
}

// Function to check if branch exists
function branchExists(branchName) {
  try {
    execSync(`git show-ref --verify --quiet refs/remotes/origin/${branchName}`, {
      stdio: 'pipe'
    });
    return true;
  } catch {
    return false;
  }
}

// Function to resolve merge conflicts
function resolveConflicts() {
  console.log('🔧 Resolving merge conflicts...');
  
  // Get list of conflicted files
  const conflictedResult = safeGitCommand('git diff --name-only --diff-filter=U', 'Get conflicted files');
  if (!conflictedResult.success) {
    console.log('❌ Could not get conflicted files');
    return false;
  }
  
  const conflictedFiles = conflictedResult.result
    .split('\n')
    .filter(f => f.trim())
    .map(f => f.trim());
  
  if (conflictedFiles.length === 0) {
    console.log('✅ No conflicted files found');
    return true;
  }
  
  console.log(`📁 Found ${conflictedFiles.length} conflicted files`);
  
  // Strategy 1: Try to resolve by accepting theirs (incoming changes)
  console.log('🔄 Attempting to resolve conflicts by accepting incoming changes...');
  const resolveResult = safeGitCommand('git checkout --theirs .', 'Accept incoming changes');
  
  if (resolveResult.success) {
    const addResult = safeGitCommand('git add .', 'Add resolved files');
    if (addResult.success) {
      console.log('✅ Conflicts resolved successfully');
      return true;
    }
  }
  
  // Strategy 2: Try to resolve by accepting ours (current changes)
  console.log('🔄 Attempting to resolve conflicts by accepting current changes...');
  const resolveOursResult = safeGitCommand('git checkout --ours .', 'Accept current changes');
  
  if (resolveOursResult.success) {
    const addOursResult = safeGitCommand('git add .', 'Add resolved files');
    if (addOursResult.success) {
      console.log('✅ Conflicts resolved successfully');
      return true;
    }
  }
  
  console.log('❌ Failed to resolve conflicts automatically');
  return false;
}

// Function to merge a single branch
function mergeBranch(branchName, index, total) {
  console.log(`\n${'='.repeat(80)}`);
  console.log(`🔄 [${index}/${total}] Processing branch: ${branchName}`);
  console.log(`${'='.repeat(80)}`);
  
  // Check if branch exists
  if (!branchExists(branchName)) {
    console.log(`❌ Branch ${branchName} does not exist`);
    return { success: false, reason: 'branch_not_found' };
  }
  
  // Fetch the branch
  console.log(`📥 Fetching branch ${branchName}...`);
  const fetchResult = safeGitCommand(`git fetch origin ${branchName}`, `Fetch ${branchName}`);
  if (!fetchResult.success) {
    console.log(`❌ Failed to fetch branch ${branchName}`);
    return { success: false, reason: 'fetch_failed' };
  }
  
  // Try to merge
  console.log(`🔀 Attempting to merge ${branchName} into main...`);
  const mergeResult = safeGitCommand(
    `git merge origin/${branchName} --no-ff -m "Merge ${branchName} into main"`,
    `Merge ${branchName}`
  );
  
  if (mergeResult.success) {
    console.log(`✅ Successfully merged ${branchName}`);
    return { success: true, method: 'direct' };
  }
  
  // Check for conflicts
  const statusResult = safeGitCommand('git status --porcelain', 'Check git status');
  if (statusResult.success && statusResult.result.includes('UU')) {
    console.log(`⚠️  Merge conflicts detected for ${branchName}`);
    
    // Try to resolve conflicts
    if (resolveConflicts()) {
      // Complete the merge
      const commitResult = safeGitCommand(
        `git commit --no-edit -m "Merge ${branchName} into main (conflicts resolved)"`,
        `Complete merge for ${branchName}`
      );
      
      if (commitResult.success) {
        console.log(`✅ Merged ${branchName} with conflict resolution`);
        return { success: true, method: 'conflict_resolved' };
      } else {
        console.log(`❌ Failed to complete merge for ${branchName}`);
        safeGitCommand('git merge --abort', `Abort merge for ${branchName}`);
        return { success: false, reason: 'commit_failed' };
      }
    } else {
      console.log(`❌ Failed to resolve conflicts for ${branchName}`);
      safeGitCommand('git merge --abort', `Abort merge for ${branchName}`);
      return { success: false, reason: 'conflict_resolution_failed' };
    }
  } else {
    console.log(`❌ Failed to merge ${branchName} (unknown error)`);
    safeGitCommand('git merge --abort', `Abort merge for ${branchName}`);
    return { success: false, reason: 'merge_failed' };
  }
}

// Main execution function
async function main() {
  console.log(`🚀 Starting Comprehensive Merge Process at ${new Date().toISOString()}`);
  console.log(`${'='.repeat(80)}\n`);
  
  // Ensure we're on main branch
  console.log('📍 Ensuring we\'re on main branch...');
  const checkoutResult = safeGitCommand('git checkout main', 'Switch to main branch');
  if (!checkoutResult.success) {
    console.log('❌ Failed to checkout main branch');
    process.exit(1);
  }
  
  // Pull latest changes
  console.log('📥 Pulling latest changes from main...');
  const pullResult = safeGitCommand('git pull origin main', 'Pull latest changes');
  if (!pullResult.success) {
    console.log('⚠️  Warning: Failed to pull latest changes, continuing anyway...');
  }
  
  // Get all unmerged branches
  console.log('📋 Getting unmerged branches...');
  const branchesResult = safeGitCommand('git branch -r --no-merged origin/main', 'Get unmerged branches');
  if (!branchesResult.success) {
    console.log('❌ Failed to get unmerged branches');
    process.exit(1);
  }
  
  const allBranches = branchesResult.result
    .split('\n')
    .filter(branch => branch.trim())
    .map(branch => branch.trim().replace('origin/', ''))
    .filter(branch => branch !== 'HEAD')
    .filter(branch => !branch.includes('backup'))
    .filter(branch => !branch.includes('disabled'));
  
  console.log(`Found ${allBranches.length} unmerged branches\n`);
  
  // Track results
  const results = {
    successful: [],
    failed: [],
    summary: {
      total: allBranches.length,
      successful: 0,
      failed: 0,
      methods: {
        direct: 0,
        conflict_resolved: 0,
        failed: 0
      }
    }
  };
  
  // Process branches in batches to avoid overwhelming the system
  const batchSize = 5;
  const batches = [];
  for (let i = 0; i < allBranches.length; i += batchSize) {
    batches.push(allBranches.slice(i, i + batchSize));
  }
  
  console.log(`📦 Processing ${batches.length} batches of up to ${batchSize} branches each\n`);
  
  // Process each batch
  for (let batchIndex = 0; batchIndex < batches.length; batchIndex++) {
    const batch = batches[batchIndex];
    console.log(`\n🔄 Processing batch ${batchIndex + 1}/${batches.length} (${batch.length} branches)...`);
    
    for (const branch of batch) {
      const result = mergeBranch(branch, results.successful.length + results.failed.length + 1, allBranches.length);
      
      if (result.success) {
        results.successful.push({ branch, ...result });
        results.summary.successful++;
        results.summary.methods[result.method]++;
      } else {
        results.failed.push({ branch, ...result });
        results.summary.failed++;
        results.summary.methods.failed++;
      }
    }
    
    // Small delay between batches
    if (batchIndex < batches.length - 1) {
      console.log('⏳ Waiting 2 seconds before next batch...');
      await new Promise(resolve => setTimeout(resolve, 2000));
    }
  }
  
  // Final summary
  console.log(`\n${'='.repeat(80)}`);
  console.log('📊 MERGE SUMMARY');
  console.log(`${'='.repeat(80)}`);
  console.log(`Total branches: ${results.summary.total}`);
  console.log(`✅ Successfully merged: ${results.summary.successful}`);
  console.log(`❌ Failed to merge: ${results.summary.failed}`);
  console.log(`Success rate: ${Math.round((results.summary.successful / results.summary.total) * 100)}%`);
  
  if (results.successful.length > 0) {
    console.log(`\n✅ Successfully merged branches:`);
    results.successful.forEach((result, index) => {
      console.log(`  ${index + 1}. ${result.branch} (${result.method})`);
    });
  }
  
  if (results.failed.length > 0) {
    console.log(`\n❌ Failed branches:`);
    results.failed.forEach((result, index) => {
      console.log(`  ${index + 1}. ${result.branch} (${result.reason})`);
    });
  }
  
  // Run system checks
  console.log(`\n🔍 Running system checks...`);
  const typeCheck = safeGitCommand('npm run type-check', 'TypeScript type checking');
  const buildCheck = safeGitCommand('npm run build', 'Production build');
  const testCheck = safeGitCommand('npm test', 'Run tests');
  
  // Push changes if any were merged
  if (results.summary.successful > 0) {
    console.log(`\n📤 Pushing merged changes to origin/main...`);
    const pushResult = safeGitCommand('git push origin main', 'Push changes to main');
    if (pushResult.success) {
      console.log('✅ Successfully pushed all changes to main!');
    } else {
      console.log('❌ Failed to push changes. You may need to push manually.');
      console.log('Run: git push origin main');
    }
  }
  
  // Generate comprehensive report
  const finalReport = {
    timestamp: new Date().toISOString(),
    summary: results.summary,
    systemChecks: {
      typeCheck: typeCheck.success,
      buildCheck: buildCheck.success,
      testCheck: testCheck.success,
      allPassed: typeCheck.success && buildCheck.success && testCheck.success
    },
    successful: results.successful,
    failed: results.failed,
    status: results.summary.successful > 0 ? 'success' : 'no-changes'
  };
  
  // Save detailed report
  fs.writeFileSync('comprehensive-merge-report.json', JSON.stringify(finalReport, null, 2));
  console.log(`\n📝 Detailed report saved to: comprehensive-merge-report.json`);
  
  console.log(`\n🏁 Merge process completed at ${new Date().toISOString()}`);
  
  return results.summary.failed === 0 ? 0 : 1;
}

// Run the main function
main().catch(error => {
  console.error(`\n💥 Fatal error: ${error}`);
  process.exit(1);
});